assayist/client/query.py

```diff
@@ -44,3 +46,144 @@ def get_container_content_sources(build_id):
+    query = """
+    // First get all the input source locations
+    MATCH (input_sl:SourceLocation) WHERE input_sl.url IN [{0}]
+    // Then find all the source locations that are upstream or downstream of the input source
+    // locations recursively. The resulting `input_and_upstream_sl` variable has all the input
+    // source locations and all the upstream or downstream source locations of the input source
+    // locations.
+    MATCH (input_sl)-[:UPSTREAM*0..]-(input_and_upstream_sl:SourceLocation)
+    // Then find all the source locations that embed the source locations in
+    // `input_and_upstream_sl` recursively. The resulting `input_upstream_and_embedded_sl`
+    // variable will have the contents of the `input_and_upstream_sl` variable previously and
+    // all the source locations that eventually embed those source locations.
+    MATCH (input_and_upstream_sl)<-[:EMBEDS*0..]-(input_upstream_and_embedded_sl:SourceLocation)
+    RETURN ID(input_upstream_and_embedded_sl)
+    """.format(', '.join(repr(sl['url']) for sl in source_locations if 'url' in sl))
+    results, _ = neomodel.db.cypher_query(query)
+    # This should only be true if none of the input source locations are in the DB
+    if not results:
+        return []
+    all_sl_ids = set([r[0] for r in results])
+
+    # Find all the artifacts that were built from the source locations, and return the Neo4j ID
+    # of those artifacts
+    query = """
+    // First get all the input source locations
+    MATCH (sl) WHERE ID(sl) IN [{0}]
+    // Find all the artifacts that were produced from a build that was built from the source
+    // locations
+    MATCH (sl)<-[:BUILT_FROM]-(:Build)-[:PRODUCED]->(artifact:Artifact)
+    RETURN ID(artifact)
+    """.format(', '.join(str(sl_id) for sl_id in all_sl_ids))
+    results, _ = neomodel.db.cypher_query(query)
+    built_with_artifact_ids = set([r[0] for r in results])
+
+    # Find all the container images that embed an artifact built from the source locations,
+    # and return the Neo4j ID of those artifacts
+    query = """
+    // First get all the artifacts that were built from the source locations
+    MATCH (artifact) WHERE ID(artifact) IN [{0}]
+    // Find all the container images that embed an artifact built from any of the source
+    // locations
+    MATCH (artifact)<-[:EMBEDS]-(container_with_embedded_artifact:Artifact {{type: 'container'}})
+    RETURN ID(container_with_embedded_artifact)
+    """.format(', '.join(repr(artifact_id) for artifact_id in built_with_artifact_ids))
+    results, _ = neomodel.db.cypher_query(query)
+    container_with_embedded_artifact_ids = set([r[0] for r in results])
+
+    # Find all the container images that were directly built from any of the source locations,
+    # and find all the container images that are children of those container images. Then return
+    # the Neo4j IDs of those container image artifacts
```

Comment:
This query, and the previous query, seem to be asking very similar questions, given that:
* a "container image [...] directly built from any of the ousrce locations" (this query) is an "artifact built from the source locations" (previous query), and
* "container images that are children of those container images" (this query) is a "container image that embeds an artifact built from the source locations" (previous query)

It's the similarity of these two queries that is making me have to re-read them, and I still don't really get why they are distinct.

Don't we want to find all the artifacts that were built directly from the source locations, then find all the type=container artifacts that either are those artifacts or ultimately embed them? (i.e. `-[:EMBEDS*0..]->`)